### DISCUSSION TITLE
Coding Style

### DISCUSSION THREAD

**Discussion by phulstaert** [Leantime/leantime#739] [General] [answered]
Could be I missed it, but is there a coding style we want to use within this project? PSR-12 for example?
I was looking at some code and have seen inconsistencies. This makes working on the project by others not as easy as it could be...

**Comment by marcelfolaron** [comment, marked as answer]
Oh I fully agree about the inconsistencies. Over the last year I've tried
to clean things up a bit but there is a lot of code...
Andyes, it should all be PSR-12.
Biggest issue might be around namespaces and folder structure, so we might
want to touch that a little later.

Some general thoughts and source of inconsistencies:

- When Leantime started the controllers only contained a run() method to
execute any request method. Some time ago I changed that to have different
controller methods for the different request methods
Parameters are passed in (but not sanitized in any way right now) The run()
method is still in there but should be considered deprecated.

- Authentication and Authorization was hard coded pretty early on, which is
why it is taking so long to create a flexible user role system. My initial
thoughts would be to take out the authentication and move it into the
controllers (which would also allow us to have public controllers).
Additionally the authorization should be decoupled and added into the
controllers (and templates where needed) as well.

- Abstracting layouts: Once authentication moves into the controllers we
can start decoupling layouts and entry points. Right now Leantime has 3
different entry points, which basically represent the layouts:
Index->content.php : Most logged in pages
Index->login.php : Login page
Install->install.php : Install page
Download->download.php : managing file downloads
In a perfect world we would have 3 layout files: default, twoColumn, empty.

Users could then update then create custom layout files in the future.

- API module: I'd like to be able to offer a restful api, so the API module
is growing and should be thought of as "the" restful api. All responses in
there should be json. Leantime is already using all of the endpoints in
there to offer more asynchronous processing. Once auth is implemented
correctly we can start offering the API.

There is probably more but those are some high level thoughts.

Thank you for asking this question!

**Reply by phulstaert** [reply]
Is it an idea to make a changelog and a roadmap file?
This would encourage other people to work on the right things and in the right order.
Also, a file with something like an ERD and/or flows would be nice. Although this takes a fair bit of time to make, it would greatly improve the ease with one would be able to start developing.

**Reply by marcelfolaron** [reply]
The changelog is available via the github releases or via tag comparisons: https://github.com/Leantime/leantime/compare/v2.1.8...v2.1.9

I think we should keep the roadmap in the discussion forum. It's been difficult to time my work but I should be able to suggest a stack rank of features that we can all discuss

**Reply by gloriafolaron** [reply]
@phulstaert the roadmap is a great thought.  With @marcelfolaron wanting to keep it as a discussion, I’d be happy to get something going in a thread and go back through the feature requests to prioritize some things.

From your questioning, I would guess that you’re looking to contribute but may be finding it hard to find a place to start? Or do you have some specific features in mind you’d want to tackle in this too? 

Thanks for adding in your thoughts — brings up some good points.

**Reply by phulstaert** [reply]
I do want to contribute, but it has been a good few years since I did some programming. Nowadays, I do more project management. Although, at home I "dabble" a bit in php/sql/docker/alpine/nginx/..., but things have changed rapidly the last few years.

On the other hand, I am not afraid to do some tedious or repetitive work if it would help the project.
